test/ca_updatedb.c

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifdef OPENSSL_SYS_UNIX
+# include <unistd.h>
+#endif
+#include "../apps/include/ca.h"
+#include "../apps/include/ca_logic.h"
+#include "../apps/include/apps.h"
+
+/* tweaks needed for Windows */
+#ifdef _WIN32
+# define access _access
+# define timezone _timezone
+#endif
+
+char *default_config_file = NULL;
+
+int main(int argc, char *argv[])
```

Comment:
I suggest using the test framework.. (libtestutil.a)

#include "testutil.h"
See most other test files which use setup_tests(), cleanup_tests(), test_get_options().

And use opt_next() to retrieve the options..

Then use the TEST_ macros e.g, TEST_true() & TEST_ptr() each test then just returns 0 or 1.

Reply:
I've never used that, so please forgive if my questions look trivial:
If I understand this correctly, this would move the whole testing from `test/recipes/80-test_ca_updatedb.t` into `test/ca_updatedb.c`. Correct?

Reply:
The test system works using  .t file(s) which can run your exe..
If you link in libtestutil.a it already has a main (see testutil/main)

test/README-dev.md has some info on this.

Look at some other test .c files to see their usage.

Reply:
It is up to you how the .t file and the exe interact.

Reply:
I tried that, but I'm a stuck - perhaps you have an idea. My issue is a linker error:
```/usr/bin/ld: test/libtestutil.a(libtestutil-lib-options.o): in function `opt_printf_stderr':
options.c:(.text+0x170): multiple definition of `opt_printf_stderr'; apps/libapps.a(libapps-lib-apps.o):apps.c:(.text+0x51a0): first defined here```
I looked and found `int opt_printf_stderr(const char *fmt, ...)` in `apps/lib/apps.c` and in `test/testutil/options.c`.
I need to include `apps.h` because that's where `load_index` is defined (and some additional functions).
What I understand least: this is caused by following line:
`testdateutc = asn1_string_to_time_t(testdate);`
When I comment the line, the linker has no issue (of course, the resulting binary doesn't make sense).
The linker error even occurs, when I change the function `asn1_string_to_time_t` to only `return NULL`.

What I generally did:
- replace `int main(int argc, char *argv[])`by `int setup_tests(void)`
- use `test_get_argument_count()` to get the amount of parameters
- replace `argv[n]` by retrieving the values through `test_get_argument(n)`

Reply:
It looks like you are trying to include the entire libapps.a which you should not do.
The whole point of this was to separate out the bits of the apps that you need into separate source files and only use them in the test/build.info.. So it should look something like this..

````
  SOURCE[ca_updateddb]=ca_updateddb.c  ../apps/ca_helpers.c ./apps/asn1_helpers.c
  INCLUDE[ca_updateddb]=.. ../include ../apps/include
  DEPEND[ca_updateddb]=../libcrypto.a libtestutil.a
````

Reply:
When you start pulling in libapps any code you pull in will likely pull in other depandancies.
The test system already has its own copy of the options which is why you are getting a clash.

Reply:
So the solution is to create an additional library (for example libca.a) and move all functions I need from apps to this? I would have to move some functions there I have not touched until now (load_index, save_index, rotate_index and probably some others)?

Reply:
You dont need to make a library .. just compile the few files you need from apps as shown in the build.info above.